### ISSUE TITLE
Add other means of identifying a service

### ISSUE THREAD

**Issue by RodrigoLeiteF** [web-scrobbler/web-scrobbler#1882]
Is it possible to add an option for connectors to identify a service through means other than the website's URL? It would make adding support for self-hosted platforms such as https://funkwhale.audio, which are not tied to a specific pattern or URL, but work consistently across all of them.

**Comment by inverse** [comment]
We provide the ability to change the URL for a given connector. See https://github.com/web-scrobbler/web-scrobbler/wiki/Custom-URL-patterns

**Comment by RodrigoLeiteF** [comment]
It works in the short term, but I think ideally it's best to have something like having the extension look for certain meta tags to automatically identify such services.

**Comment by EliotBerriot** [comment]
Hi, Funkwhale maintainer here :)

If needed, I can easily add a meta tag into Funkwhale HTML, like `<meta name="generator" content="funkwhale 0.18">`. I'm not sure how the actual scrobbling work though. How does Web Scrobbler identifies that the player switched to another track? If I can help with that, let me know!

**Comment by EliotBerriot** [comment]
> The scrobbling normally works by binding to DOM elements and watching for changes. e.g. like this: https://github.com/web-scrobbler/web-scrobbler/blob/master/src/connectors/mnet.js

Makes sense, I'll try to keep stable class names for artists / albums / track data then!

**Comment by alexesprit** [comment]
I've added a basic support for identifying websites via checking for match rules in a content script. Will merge these changes into master branch after some testing and extending the match rules.

**Comment by alexesprit** [comment]
@EliotBerriot would be great if you add a meta tag like you suggested.  I think it should contain no version:
```xml
<meta name="generator" content="Funkwhale">
```
as the version is redundant here (can be added as a separate `meta` tag, though).

**Comment by EliotBerriot** [comment]
@alexesprit, sure, I'll do that today, and it will be included in our next release :)

**Comment by EliotBerriot** [comment]
@alexesprit So it's been implemented, and deployed at https://open.audio/library if you want to try it. Let me know if you need me to set some static classes / attributes on the player to make the job easier for you :)

**Comment by stephanebruckert** [comment]
Can we find a method to do this with the Spotify and Bandcamp embeds as well? These are still open/unmerged:
 
- https://github.com/web-scrobbler/web-scrobbler/issues/2024
- https://github.com/web-scrobbler/web-scrobbler/pull/2026

**Comment by alexesprit** [comment]
@stephanebruckert the way I added should work for embeds in theory, but I haven't checked it.

There're some known limitations:
- It doesn't work in websites that add embeds dynamically (e.g. reddit, where a post with an embed can be loaded while scrolling posts).
- It doesn't work in websites that already covered by another connector.

**Comment by inverse** [comment]
@alexesprit did you make any progress on this injection mechanism?

**Comment by alexesprit** [comment]
https://github.com/web-scrobbler/web-scrobbler/tree/feature/request-inject but it's not finished.
